### DISCUSSION TITLE
What's going on with InvoicePlane + PHP 8.2

### DISCUSSION THREAD

**Discussion by nielsdrost7** [InvoicePlane/InvoicePlane#1067] [Future Versions]
_Originally posted by @Masgalor in https://github.com/InvoicePlane/InvoicePlane/issues/1063#issuecomment-2252477286_

In [one of our issues](https://github.com/InvoicePlane/InvoicePlane/issues/1063) we had a great question:

    I am looking for the best option to run IP on PHP8.2.

Here's the quote from that issue:

> But thanks for your answer anyway, it might be helpful for others.
> 
> Since Debian-stable ships PHP8.2 I have no way to use something below this, what can I do to run IP?
> If I got it right I can either use the information from https://github.com/InvoicePlane/InvoicePlane/issues/1009#issuecomment-1962227003 to build IPv1 myself with an unofficial fork of CI3 (that supports PHP8.2) or try IPv2 which is still in development.
> 
> But this raises the next question, where can I find IPv2? Google gives me results of people talking about fit or at least 6 years, but I can not find any repository or download or any other signs of it's existence.

And of course:

    So, if I really need to run IP on PHP8.2 what is the best thing I can do?

**Comment by nielsdrost7** [comment]
Hey @Masgalor I've created a discussion for you so I can answer your questions in a better way.

**Reply by nielsdrost7** [reply]
> I am looking for the best option to run IP on PHP8.2.

IP & PHP 8.2:

(feel free to ask questions about below steps, if you like)

InvoicePlane was build with something that's called `CodeIgniter version 3`.

1. **CodeIgniter 3** (after this: **CI3**) has 2 things:
  a. A branch with an almost working version for them for PHP 8.2,
  b. Someone who couldn't wait and just fixed it.

In your ` composer.json` point to either of those 2 solutions, then run `composer update`

3. IP: Turn on any kind of debug you can find. `error_reporting`, everything. Now load IP on your PHP 8.2 with `CI3` solution

4. You'll find a couple of similar cases: `assignment of Dynamic Properties`. Long story, but you'll find them in your error messages.

I don't know if that's all the errors you'll find, but it's the thing that changed between PHP `8.1` and `8.2`

Once the `dynamic properties` situation has been resolved, the rest should be fairly easy to fix (_famous last words_)

Hope it helps

**Reply by Masgalor** [reply]
Thank your very much for your effort, I will have a look next week.

>a. A branch with an almost working version for them for PHP 8.2,

The PR in their repo (potentially) resolving this is almost 2 years old and looks kind of stale to me.
I can understand why someone fixed it in a fork.

**Reply by nielsdrost7** [reply]
If you really want to go with PHP 8.2 then I would go for that fork.
Only thing you need to do then is fix those problems with `assignment of Dynamic Properties` (if IP has any)

Are you going to try it out _locally_ or on your server?

**Reply by Masgalor** [reply]
I will put on my existing server structure.
I am not long at the place/PC during the day so I can access it from everywhere.

**Reply by nielsdrost7** [reply]
> I will put on my existing server structure. I am not long at the place/PC during the day so I can access it from everywhere.

Expect lots of errors in the beginning. I would have turned on `error_reporting(E_ALL);` in my php.ini
Now with that turned on it (hopefully) shows errors and where they come from.
Let's say IPv1 is installed on https://invoicing.underdog.com

If someone would access that and saw those error messages, they would get information about your server.

I think even on `nginx` you can make some kind of .htpassword popup to protect you from that

**Reply by nielsdrost7** [reply]
PHPStorm has an inspection that runs through all files of InvoicePlane and detects the `assignment of Dynamic Properties` problem.

It also has an evaluation subscription for 1 month

**Reply by nielsdrost7** [reply]
https://github.com/mhzawadi/invoiceplane/tree/dev (the `dev` branch) has PHP 8.2 on Docker
`docker-compose up` will let you run InvoicePlane locally on your PC
Run that inspection with PHPStorm (if there's another name for it, I'll let you know)

All those files with that problem need to be adjusted. After that I would:
- Try it out locally, see if there are more errors
- Then put the corrected file on the server and run InvoicePlane on your server
